[EBPF] fix flaky gpu-module test

val06 opened the pull request:
### What does this PR do?

Fixes the flaky test (TestDetectsContainer) by switching to use pkg/util/testutil/docker (see this https://github.com/DataDog/datadog-agent/pull/31470) and making the cudasample process to wait until killed explicitly (avoiding potential race conditions when trying to get docker's main process PID)

### Motivation

fix race condition when trying to detect container's main pid. was returning 0, if the container finished its execution following its main process exit

### Describe how to test/QA your changes

all /pkg/gpu pkg tests should pass.
Need to monitor on main that `TestDetectsContainer` is not flaky anymore

### Possible Drawbacks / Trade-offs

### Additional Notes

[Jira ticket](https://datadoghq.atlassian.net/browse/EBPF-620)

val06 committed:
removed endWaitTime arg from the cudasample and changed the process to wait for input instead of exiting

val06 committed:
- refactored docker tests to use the dockerutils
- changed some structs to be private
- added use of PatternScanner to validate successful invocation of the cudasample binary

val06 committed:
removed the flaky marker from TestDetectsContainer

val06 committed:
;

val06 committed:
fixed cudasample help message

val06 committed:
fixed go linter

val06 committed:
added cmd.wait to cleanup when running the bare process

val06 committed:
fixed another linter error

val06 committed:
changed to use finishedPattern to track end of execution for cudasample

val06 committed:
changed cudasample to use pause() instead of getchar()

val06 committed:
killing target program
get pid,cid without require.Eventually (since we have the patternScanner to validate proper execution)

val06 committed:
added comment to the printed line used as finishedPattern marker

val06 committed:
small fixes

val06 committed:
returned utils.WaitForProgramsToBeTraced to ensure the cudasample was detected by the uprobeattacher

gjulianm commented on pkg/gpu/probe_test.go:
```diff
@@ -145,21 +142,17 @@ func (s *probeTestSuite) TestMultiGPUSupport() {
-
-	// Wait until the process finishes and we can get the stats. Run this instead of waiting for the process to finish
-	// so that we can time out correctly
-	require.Eventually(t, func() bool {
-		return !utils.IsProgramTraced(gpuModuleName, gpuAttacherName, cmd.Process.Pid)
-	}, 60*time.Second, 500*time.Millisecond, "process not stopped")
```
I'd leave this part just to have a sanity check that the gpu probe recognized that the process is ended. We could reduce the timeout to something more reasonable (500 ms maybe?)

val06 replied:
fixed the other cause for flakyness.
this check is not needed anymore, since Run command ensures the target is executed until the end (by checking the finishPattern). the target process will then hang until killed (happens in the test cleanup)

gjulianm reviewed (approved):
LGTM, just two minor nitpicks

val06 committed:
removed redundant printf

val06 committed:
moved command log print only when succeeded

val06 committed:
removed redundant cudasampel kills and removed (again) unnecessary waitfortracedprogram calls

fixed flaky tests, by adding check for existing streamHandlers before trying to get the stats (due to the race with the EventConsumer)

val06 commented:
/merge
